Start replacing `TRUELENGTH` markers with a hash

aitap opened the pull request:
With apologies to Matt Dowle, who had poured a lot of effort into making `data.table` go fast.

Ongoing work towards #6180. Unfortunately, doesn't completely remove any uses of non-API entry points by itself. Detailed motivation [here](https://github.com/rdatatable-community/The-Raft/blob/56ccfaad8251839d4ed17136274f67abe4c3924a/posts/2024-12-02-non-api-use/index.qmd#L348-L407) in a pending blog post. Can't start implementing [stretchy ALTREP vectors](https://github.com/rdatatable-community/The-Raft/blob/56ccfaad8251839d4ed17136274f67abe4c3924a/posts/2024-12-02-non-api-use/index.qmd#L307-L344) until `data.table` stops using `TRUELENGTH` to mark them.

Currently implemented:

- An open-addressing hash table with Knuth's linear multiplication hashing
  - Deliberately no removal operation, only insertion and update
  - Can be made more performant (collisions much less likely) by switching to tables of power-of-two length and double hashing
- Almost all uses of `TRUELENGTH` to mark `CHARSXP`s or columns replaced with a hash

Needs more work:
- [ ] one use in `rbindlist()` pre-allocates a lot of unneeded memory
  - will probably require a dynamically growing hash table
- [ ] `forder.c`, the last remaining user of `savetl`
  - `SET_TRUELENGTH` is atomic, `hash_set` is not, will need additional care in multi-threaded environment
- [ ] `savetl` machinery in `assign.c`

Let's just see how much worse is the performance going to get.

tdhock commented:
hi, thanks for this. Can you please propose one or two performance test cases that you think may be adversely affected by these changes? Is it when we create a table with one column, and then use `:=` or `set` to add a lot more columns?

aitap commented:
The `rbindlist()` code pre-allocates `O(sum(lengths(lapply(l, names)))` memory. In the best case, only `O(length(names(l[[1]])))` is needed, `length(l)` times less. In the worst case, all names are different and the code needs all that memory for `fill = TRUE`.

The `forder()` code pre-allocates `O(nrow(x))` memory.

I'll try giving many `data.table`s to `rbindlist()` and long `data.table`s with string columns to `setorder()`.

SebKrantz commented:
Perhaps you can also try a fast 3rd party hash map: https://martin.ankerl.com/2019/04/01/hashmap-benchmarks-01-overview/

I particular Google's Abseil hash is pretty fast: https://abseil.io/docs/cpp/guides/container https://abseil.io/docs/cpp/guides/hash

aitap commented:
It's pretty bad. For typical cases, the current hash table eats >1 order of magnitude more memory, and it's similarly slower in `forder()`:
![forder, typical case](https://github.com/user-attachments/assets/2a3a3429-1b96-4d21-b857-a15dc8c25f78)
![rbindlist, typical case](https://github.com/user-attachments/assets/4342a591-acd9-495a-91db-003323d19f97)

<details>

The hash table is only on par by time in the worst case for `forder()` (all strings different) and might be on par in the worst case for `rbindlist(fill = TRUE)` (all column names different), but the latter is completely unrealistic as a use case.

![forder, worst case](https://github.com/user-attachments/assets/952796ac-5439-4a2a-bbc8-b4b19c3e73f6)
![rbindlist, worst case](https://github.com/user-attachments/assets/115fa1f8-c317-452a-9888-65a37ee02050)

```r
```

tdhock commented:
thanks for proposing the performance test cases and sharing the atime benchmark graphs. I agree that we should try to avoid an order of magnitude constant factor increase in time/memory usage.

aitap commented:
Since profiling has shown that a noticeable amount of time is wasted initialising the giant pre-allocated hash tables, I was able to make the slowdown factor closer to 2 by dynamically re-allocating the hash table:
![dynamic_hash](https://github.com/user-attachments/assets/63c6f9ac-7f70-4d03-bc14-1464d5a85e20)

The memory use is significantly reduced (except for the worst cases), but cannot be measured with `atime` (just like other uses of `calloc`/`malloc`/`realloc` in `data.table` are invisible to `Rprofmem`).

<details>

```r
library(atime)

pkg.path <- '.'
limit <- 1
# taken from .ci/atime/tests.R
pkg.edit.fun <- function(old.Package, new.Package, sha, new.pkg.path) {
  pkg_find_replace <- function(glob, FIND, REPLACE) {
    atime::glob_find_replace(file.path(new.pkg.path, glob), FIND, REPLACE)
  }
  Package_regex <- gsub(".", "_?", old.Package, fixed = TRUE)
  Package_ <- gsub(".", "_", old.Package, fixed = TRUE)
  new.Package_ <- paste0(Package_, "_", sha)
  pkg_find_replace(
    "DESCRIPTION",
    paste0("Package:\\s+", old.Package),
    paste("Package:", new.Package))
  pkg_find_replace(
    file.path("src", "Makevars.*in"),
    Package_regex,
    new.Package_)
  pkg_find_replace(
    file.path("R", "onLoad.R"),
    Package_regex,
    new.Package_)
  pkg_find_replace(
    file.path("R", "onLoad.R"),
    sprintf('packageVersion\\("%s"\\)', old.Package),
    sprintf('packageVersion\\("%s"\\)', new.Package))
  pkg_find_replace(
    file.path("src", "init.c"),
    paste0("R_init_", Package_regex),
    paste0("R_init_", gsub("[.]", "_", new.Package_)))
  pkg_find_replace(
    "NAMESPACE",
    sprintf('useDynLib\\("?%s"?', Package_regex),
    paste0('useDynLib(', new.Package_))
}

versions <- c(
  master   = '70c64ac08c6becae5847cd59ab1efcb4c46437ac',
  static_hash = '24e81785669e70caac31501bf4424ba14dbc90f9',
  dynamic_hash = 'd7a9a1707ec94ec4f2bd86a5dfb5609207029ba4'
)

N <- 10^seq(2, 7.5, .25)
# expected case: a few distinct strings
forderv1_work <- lapply(setNames(nm = N), \(N)
  sample(letters, N, TRUE)
)
forderv1 <- atime_versions(
  pkg.path, N,
  expr = data.table:::forderv(forderv1_work[[as.character(N)]]),
  sha.vec = versions, seconds.limit = limit, verbose = TRUE,
  pkg.edit.fun = pkg.edit.fun
)
rm(forderv1_work); gc(full = TRUE)

# worst case: all strings different
# (a challenge for the allocator too due to many small immovable objects)
N <- 10^seq(2, 7.5, .25)
forderv2_work <- lapply(setNames(nm = N), \(N)
  format(runif(N), digits = 16)
)
forderv2 <- atime_versions(
  pkg.path, N,
  expr = data.table:::forderv(forderv2_work[[as.character(N)]]),
  sha.vec = versions, seconds.limit = limit, verbose = TRUE,
  pkg.edit.fun = pkg.edit.fun
)
rm(forderv2_work); gc(full = TRUE)

# expected case: all columns named the same
N <- 10^seq(1, 5.5, .25) # number of data.tables in the list
k <- 10 # number of columns per data.table
rbindlist1_work <- lapply(setNames(nm = N), \(N)
  rep(list(setNames(as.list(1:k), letters[1:k])), N)
)
rbindlist1 <- atime_versions(
  pkg.path, N,
  expr = data.table::rbindlist(rbindlist1_work[[as.character(N)]]),
  sha.vec = versions, seconds.limit = limit, verbose = TRUE,
  pkg.edit.fun = pkg.edit.fun
)
rm(rbindlist1_work); gc(full = TRUE)

# worst case: all columns different
N <- 10^seq(1, 4.5, .25) # number of data.tables in the list
k <- 10 # number of columns per data.table
rbindlist2_work <- lapply(setNames(nm = N), \(N)
  replicate(N, setNames(as.list(1:k), format(runif(k), digits = 16)), FALSE)
)
rbindlist2 <- atime_versions(
  pkg.path, N,
  expr = data.table::rbindlist(rbindlist2_work[[as.character(N)]], fill = TRUE),
  sha.vec = versions, seconds.limit = limit, verbose = TRUE,
  pkg.edit.fun = pkg.edit.fun
)
rm(rbindlist2_work); gc(full = TRUE)

#save(forderv1, forderv2, rbindlist1, rbindlist2, file = 'times.rda')
```

</details>

The main problem with the current approach is that since the parallel loop in `src/forder.c:range_str(...)` needs to read the hash table from outside the critical section while hash elements are being set inside the critical section, it's possible that `dhash_set()` will need to reallocate the table while `dhash_lookup()` is using it. I have already tried using [pthread R/W locks](https://pubs.opengroup.org/onlinepubs/9799919799/functions/pthread_rwlock_unlock.html) and OpenMP critical sections; they were too slow. I wasn't able to use OpenMP atomics to make atomic reads/writes through a pointer (maybe that's asking for too much). Any third-party hash table will also need to be outfitted with appropriate locking.

The current code keeps one previous hash table until the next reallocation cycle and hopes that `dhash_enlarge` writes the new table pointer into the object _roughly_ at the same time as the metadata, so that when `dhash_lookup` reads the object, it will _hopefully_ receive a consistent view of the object. This is not good enough for production because (1) the race window still exists and (2) there's no proof that there _won't_ be a stale reader still using the previous table array when the current writer deallocates it. What can be done with it:

 - Don't deallocate the previous hash tables until the function is completely done with them. This will double the memory use (because the code doubles the buffer when re-allocating it and $\sum_{i=0}^{n} 2^i = 2^{n+1} - 1$). We'll be able to switch back from `malloc` to `R_alloc` and make the allocations visible to R's allocator.
 - Devise a [read-copy-update](https://en.wikipedia.org/wiki/Read-copy-update#Analogy_with_reader%E2%80%93writer_locking) scheme (see also: [1](https://github.com/nicowilliams/ctp/blob/master/thread_safe_global.c) [2](http://liburcu.org/)) to properly expire the previous versions of the hash table. These may require synchronisation primitives not easily available through OpenMP. Reference-counting the hash tables may result in contention and slowdowns similar to pthread R/W locks.

SebKrantz commented:
In case it helps, {collapse}'s hash functions (https://github.com/SebKrantz/collapse/blob/master/src/kit_dup.c and https://github.com/SebKrantz/collapse/blob/master/src/match.c) are pretty fast as well - inspired by base R -> multiplication hash using unsigned int. It's bloody fast but requires a large table. But Calloc() is quite efficient. Anyway, would be great if you'd test the Google Hash function, curious to see it it can do much better.

aitap commented:
The abseil hash function is very slightly slower in my tests, although the difference may be not significant. Perhaps that's because [my C port](https://github.com/Rdatatable/data.table/blob/159e1d48926b72af9f212b8c645a8bc8ab6b20be/src/hash.c#L5-L62) fails to inline some of the things that naturally inline in [the original C++ with templates](https://github.com/abseil/abseil-cpp/blob/65b7da13bf05228cfffc0db7c72f8343866b9c57/absl/hash/internal/hash.h#L1066-L1077). I can try harder, but that's a lot of extra code to bring in properly.

![forder() performance on a load with a few unique strings](https://github.com/user-attachments/assets/b76cebad-fc22-4af7-bf06-a2e84f8e54d2)

![forder() performance on a pathological load with all strings unique](https://github.com/user-attachments/assets/a8c46976-dfaa-4f20-82df-50320cbd7860)

<details>

```r
library(atime)

pkg.path <- '.'
limit <- 1
# taken from .ci/atime/tests.R
pkg.edit.fun <- function(old.Package, new.Package, sha, new.pkg.path) {
  pkg_find_replace <- function(glob, FIND, REPLACE) {
    atime::glob_find_replace(file.path(new.pkg.path, glob), FIND, REPLACE)
  }
  Package_regex <- gsub(".", "_?", old.Package, fixed = TRUE)
  Package_ <- gsub(".", "_", old.Package, fixed = TRUE)
  new.Package_ <- paste0(Package_, "_", sha)
  pkg_find_replace(
    "DESCRIPTION",
    paste0("Package:\\s+", old.Package),
    paste("Package:", new.Package))
  pkg_find_replace(
    file.path("src", "Makevars.*in"),
    Package_regex,
    new.Package_)
  pkg_find_replace(
    file.path("R", "onLoad.R"),
    Package_regex,
    new.Package_)
  pkg_find_replace(
    file.path("R", "onLoad.R"),
    sprintf('packageVersion\\("%s"\\)', old.Package),
    sprintf('packageVersion\\("%s"\\)', new.Package))
  pkg_find_replace(
    file.path("src", "init.c"),
    paste0("R_init_", Package_regex),
    paste0("R_init_", gsub("[.]", "_", new.Package_)))
  pkg_find_replace(
    "NAMESPACE",
    sprintf('useDynLib\\("?%s"?', Package_regex),
    paste0('useDynLib(', new.Package_))
}

versions <- c(
  master = '70c64ac08c6becae5847cd59ab1efcb4c46437ac',
  'Knuth_hash' = 'd7a9a1707ec94ec4f2bd86a5dfb5609207029ba4',
  'abseil_hash' = '159e1d48926b72af9f212b8c645a8bc8ab6b20be'
)

N <- 10^seq(2, 7.5, .25)
# expected case: a few distinct strings
forderv1_work <- lapply(setNames(nm = N), \(N)
  sample(letters, N, TRUE)
)
forderv1 <- atime_versions(
  pkg.path, N,
  expr = data.table:::forderv(forderv1_work[[as.character(N)]]),
  sha.vec = versions, seconds.limit = limit, verbose = TRUE,
  pkg.edit.fun = pkg.edit.fun
)
rm(forderv1_work); gc(full = TRUE)

# worst case: all strings different
# (a challenge for the allocator too due to many small immovable objects)
N <- 10^seq(2, 7.5, .25)
forderv2_work <- lapply(setNames(nm = N), \(N)
  format(runif(N), digits = 16)
)
forderv2 <- atime_versions(
  pkg.path, N,
  expr = data.table:::forderv(forderv2_work[[as.character(N)]]),
  sha.vec = versions, seconds.limit = limit, verbose = TRUE,
  pkg.edit.fun = pkg.edit.fun
)
rm(forderv2_work); gc(full = TRUE)

# expected case: all columns named the same
N <- 10^seq(1, 5.5, .25) # number of data.tables in the list
k <- 10 # number of columns per data.table
rbindlist1_work <- lapply(setNames(nm = N), \(N)
  rep(list(setNames(as.list(1:k), letters[1:k])), N)
)
rbindlist1 <- atime_versions(
  pkg.path, N,
  expr = data.table::rbindlist(rbindlist1_work[[as.character(N)]]),
  sha.vec = versions, seconds.limit = limit, verbose = TRUE,
  pkg.edit.fun = pkg.edit.fun
)
rm(rbindlist1_work); gc(full = TRUE)

# worst case: all columns different
N <- 10^seq(1, 4.5, .25) # number of data.tables in the list
k <- 10 # number of columns per data.table
rbindlist2_work <- lapply(setNames(nm = N), \(N)
  replicate(N, setNames(as.list(1:k), format(runif(k), digits = 16)), FALSE)
)
rbindlist2 <- atime_versions(
  pkg.path, N,
  expr = data.table::rbindlist(rbindlist2_work[[as.character(N)]], fill = TRUE),
  sha.vec = versions, seconds.limit = limit, verbose = TRUE,
  pkg.edit.fun = pkg.edit.fun
)
rm(rbindlist2_work); gc(full = TRUE)

save(forderv1, forderv2, rbindlist1, rbindlist2, file = 'times.rda')
```

</details>

`collapse::fmatch()` works very well, faster than `data.table::chmatch()` even in its current form:

![chmatch() vs. fmatch() on a best case load](https://github.com/user-attachments/assets/38f90562-9340-4746-8e43-e0dc1ef74a7c)

<details>

```r
library(atime)
library(data.table)

limit <- 1

# assumes that atime_versions() had pre-installed the packages
# master = '70c64ac08c6becae5847cd59ab1efcb4c46437ac',
# 'Knuth_hash' = 'd7a9a1707ec94ec4f2bd86a5dfb5609207029ba4',

N <- 10^seq(2, 7.5, .25)
# expected case: a few distinct strings
chmatch_work1 <- lapply(setNames(nm = N), \(N)
  sample(letters, N, TRUE)
)
chmatch1 <- atime(
  N,
  seconds.limit = limit, verbose = TRUE,
  master = data.table.70c64ac08c6becae5847cd59ab1efcb4c46437ac::chmatch(chmatch_work1[[as.character(N)]], letters),
  Knuth_hash = data.table.d7a9a1707ec94ec4f2bd86a5dfb5609207029ba4::chmatch(chmatch_work1[[as.character(N)]], letters),
  collapse = collapse::fmatch(chmatch_work1[[as.character(N)]], letters)
)
rm(chmatch_work1); gc(full = TRUE)

save(chmatch1, file = 'times_collapse.rda')
```

</details>

And the real memory cost isn't even that large:
![memory use as returned by getrusage()](https://github.com/user-attachments/assets/61bccb89-de43-4240-97ef-74fa64f4d788)

<details>

```r
library(atime)
library(data.table)
# assumes that atime_versions() had pre-installed the packages
# master = '70c64ac08c6becae5847cd59ab1efcb4c46437ac',
# 'Knuth_hash' = 'd7a9a1707ec94ec4f2bd86a5dfb5609207029ba4',
library(data.table.70c64ac08c6becae5847cd59ab1efcb4c46437ac)
library(data.table.d7a9a1707ec94ec4f2bd86a5dfb5609207029ba4)
library(parallel)

# only tested on a recent Linux system
# measures the _maximal_ amount of memory in kB used by the current process
writeLines('
#include <sys/resource.h>
void maxrss(double * kb) {
  struct rusage ru;
  int ret = getrusage(RUSAGE_SELF, &ru);
  *kb = ret ? -1 : ru.ru_maxrss;
}
', 'maxrss.c')
tools::Rcmd('SHLIB maxrss.c')
dyn.load(paste0('maxrss', .Platform$dynlib.ext))

limit <- 1

N <- 10^seq(2, 7.5, .25)
# expected case: a few distinct strings
chmatch_work1 <- lapply(setNames(nm = N), \(N)
  sample(letters, N, TRUE)
)
versions <- expression(
  master = data.table.70c64ac08c6becae5847cd59ab1efcb4c46437ac::chmatch(chmatch_work1[[as.character(N)]], letters),
  Knuth_hash = data.table.d7a9a1707ec94ec4f2bd86a5dfb5609207029ba4::chmatch(chmatch_work1[[as.character(N)]], letters),
  collapse = collapse::fmatch(chmatch_work1[[as.character(N)]], letters)
)

plan <- expand.grid(N = N, version = names(versions))
chmatch1 <- lapply(seq_len(nrow(plan)), \(i) {
  # use a disposable child process
  mccollect(mcparallel({
    eval(versions[[plan$version[[i]]]], list(N = plan$N[[i]]))
    .C('maxrss', kb = double(1))$kb
  }))
})

rm(chmatch_work1); gc(full = TRUE)

save(chmatch1, file = 'times_collapse.rda')
chmatch1p <- lattice::xyplot(
  maxrss_kb ~ N, cbind(plan, maxrss_kb = unlist(chmatch1)), group = version,
  auto.key = TRUE, scales = list(log = 10),
  par.settings=list(superpose.symbol=list(pch=19))
)
```

</details>

SebKrantz commented:
Nice! Thanks. I always wondered about this tradeoff between the size of the table and the quality of the hash function. Looks like speed + large table still wins. Anyway, if you want to adopt, feel free to copy it under your MPL license. Just mention me in the top of the file and as a contributor.

SebKrantz commented:
PS: I believe it also depends on the size of the `table` in R. `letters` is small. I believe for much larger tables `chmatch()` should be faster. Would be interesting to also check those hash functions.

tdhock commented:
excellent work thank you very much

aitap commented:
The `chmatch()` performance problem is three-dimensional.

There's `length(x)`; we'd like the lookup time to be linear in that and spend as few instructions per lookup as possible. `TRUELENGTH()` is a very short function call.

Initialising the hash currently takes time and memory proportional to `length(table)`, but we really want it to be closer to `length(unique(table))`. See the benchmark where the `table` is 10 times as large as `x`:
https://github.com/Rdatatable/data.table/blob/9fe1b8d01701d2b23adcff3085ea0290da5df022/src/chmatch.c#L163-L170

A good `TRUELENGTH()` replacement needs to perform well when `length(x)` and `length(table)` are large and `length(unique(table))` is still reasonable. (`length(unique(table))` becomes unreasonable when it starts to strain the `CHARSXP` cache.)

I still have a few ideas how to speed this up.

aitap commented on src/chmatch.c:
```diff
@@ -56,33 +55,14 @@ static SEXP chmatchMain(SEXP x, SEXP table, int nomatch, bool chin, bool chmatch
-    int tl = TRUELENGTH(s);
-    if (tl>0) { savetl(s); tl=0; }
-    if (tl==0) SET_TRUELENGTH(s, chmatchdup ? -(++nuniq) : -i-1); // first time seen this string in table
+    int tl = hash_lookup(marks, s, 0);
+    if (tl==0) hash_set(marks, s, chmatchdup ? -(++nuniq) : -i-1); // first time seen this string in table
```
`hash_set` wastes time performing the lookup again. Need to expose a "lookup or create" operation somehow to avoid that.

ben-schwen replied:
I guess we can do that analagous to `hash_set`?

```C
R_xlen_t hash_lookup_or_insert(hashtab *h, SEXP key, R_xlen_t value) {
  struct hash_pair *cell = h->tb + hash_index(key, h->multiplier) % h->size, *end = h->tb + h->size - 1;
  for (size_t i = 0; i < h->size; ++i, cell = (cell == end ? h->tb : cell + 1)) {
    if (cell->key == key) {
      return cell->value;  // Key exists, don't update, return value
    } else if (!cell->key) {
      if (!h->free) internal_error(
        __func__, "no free slots left (full size=%zu)", h->size
      );
      --h->free;
      *cell = (struct hash_pair){.key = key, .value = value};
      return value;  // insert here
    }
  }
  
  internal_error( // # nocov
    __func__, "did not find a free slot for key %p; size=%zu, free=%zu",
    (void*)key, h->size, h->free
  );
  // Should be impossible, but just in case:
  return value;
}

```

ben-schwen replied:
It can be found in 337a0c2d508a31c59885416d7929ff6d6a4b0bda. Gives a little speedup but not too much.

MichaelChirico commented:
WDYT about the possibility of proposing our %chin% just be up streamed into R itself? Our version is "well-tested & battle-hardened" for many years, and seems naturally suited to r-devel given it's reliance on internal behavior.

aitap commented on src/hash.c:
```diff
@@ -0,0 +1,226 @@
+#include "data.table.h"
+
+struct hash_pair {
+    SEXP key;
+    R_xlen_t value;
```
Can halve the memory requirements by storing `int` indices into an array instead of `SEXP` pointers and `int` mark values instead of `R_xlen_t` (thanks to Sebastian Krantz for the idea). Cost: this makes the hash table incompatible with long vectors. Since `int` dimensions for matrices and `data.frame`s are baked into almost all of R, it's probably worth using `int`s here too.

aitap commented on src/forder.c:
```diff
@@ -307,20 +302,22 @@ static void range_str(const SEXP *x, int n, uint64_t *out_min, uint64_t *out_max
-      if (TRUELENGTH(s)>0)   // save any of R's own usage of tl (assumed positive, so we can both count and save in one scan), to restore
-        savetl(s);           // afterwards. From R 2.14.0, tl is initialized to 0, prior to that it was random so this step saved too much.
-      // now save unique SEXP in ustr so i) we can loop through them afterwards and reset TRUELENGTH to 0 and ii) sort uniques when sorting too
+    // Why is it acceptable to call dhash_lookup when marks can be shared between threads?
```
So far the only way I know to make this provably safe is to throw away old hash table buffers (to be garbage collected by R).

Fast, thread-safe, dynamically growing, memory efficient hash table: choose two out of four, maybe three if you're lucky.

aitap commented:
@MichaelChirico, it may be worth a try. I suppose this would mean falling back to `TRUELENGTH` code on R < 4.6? Although the hash table by Sebastian Krantz is pretty fast and sometimes beats `data.table::chmatch` (need to explore all three performance dimensions to understand it better); we can have one like that too.

The `forder.c` problem is also quite hair-raising. We can't just replace it with `sort(method = "radix")`, can we?

The remaining uses of `TRUELENGTH` are much less performance-critical, for things that people usually don't have too many of (e.g. factor levels, columns).

jangorecki commented:
> Can't start implementing stretchy ALTREP vectors until data.table stops using TRUELENGTH to mark them.

It is worth noting that there is an ongoing effort, expected to start from May 2025 to work on R API growable vector API. Then we may expect easier, or more performant, ways to address current problems.

MichaelChirico commented:
If `%chin%` were accepted into r-devel, we could just mark `%chin%` as deprecated and meanwhile mask `` `%chin%` = `%in%`; chmatch = match ``, leaving users on older R with worse performance, WDYT

ben-schwen commented:
I extended Ivans benchmark a little further e.g. when our table gets bigger and also included NO matches. We definitely have to take a closer look at this with varying table size etc. (might also try other hashing methods like double, robin hood or cuckoo hashing)

<img width="1075" alt="2025-03-23_18-55" src="https://github.com/user-attachments/assets/ec3dc86d-f874-4fbd-8814-a9e298770edf" />


<details>

```r
library(atime)
library(data.table)

# master = '70c64ac08c6becae5847cd59ab1efcb4c46437ac',
# 'Knuth_hash' = 'd7a9a1707ec94ec4f2bd86a5dfb5609207029ba4',

sample_strings = function(N=10, len=4) {
   do.call(paste0, replicate(len, sample(LETTERS, N, TRUE), FALSE))
}

N <- 10^seq(2, 7.5, .25)
tab_full = sample_strings(1e6, 10)
tab_small = sample(tab_full, 9e5)
# expected case: a few distinct strings
chmatch_work1 <- lapply(setNames(nm = N), \(N)
  sample(tab_full, N, TRUE)
)
chmatch1 <- atime(
  N,
  seconds.limit = limit, verbose = TRUE,
  master = data.table.70c64ac08c6becae5847cd59ab1efcb4c46437ac::chmatch(chmatch_work1[[as.character(N)]], tab_small),
  Knuth_hash = data.table.d7a9a1707ec94ec4f2bd86a5dfb5609207029ba4::chmatch(chmatch_work1[[as.character(N)]], tab_small),
  collapse = collapse::fmatch(chmatch_work1[[as.character(N)]], tab_small)
)
plot(chmatch1)

rm(chmatch_work1); gc(full = TRUE)

save(chmatch1, file = 'fmatch_missings.rda')
```
</details>

HughParsonage commented on src/hash.c:
```diff
@@ -0,0 +1,141 @@
+    h = hash_rehash(h);
+  (void)hash_set_(h, key, value);
+  return h;
+}
```
Similar issue as in `hash_set`

HughParsonage commented on src/rbindlist.c:
```diff
@@ -74,7 +74,7 @@ SEXP rbindlist(SEXP l, SEXP usenamesArg, SEXP fillArg, SEXP idcolArg, SEXP ignor
     if (!uniq)
       error(_("Failed to allocate upper bound of %"PRId64" unique column names [sum(lapply(l,ncol))]"), (int64_t)upperBoundUniqueNames); // # nocov
-    savetl_init();
+    hashtab * marks = hash_create(1 + (LENGTH(l) ? (2 * LENGTH(getAttrib(VECTOR_ELT(l, 0), R_NamesSymbol))) : 0));
```
The choice to use the first name to size the hash table seems arbitrary. Is there a better way? Using `upperBoundUniqueNames`?

aitap commented on src/assign.c:
```diff
@@ -855,45 +844,41 @@ const char *memrecycle(const SEXP target, const SEXP where, const int start, con
         }
         source = newSource;
-        for (int k=0; k<nTargetLevels; ++k) SET_TRUELENGTH(targetLevelsD[k], 0);  // don't need those anymore
+        for (int k=0; k<nTargetLevels; ++k) hash_set(marks, targetLevelsD[k], 0);  // don't need those anymore
```
Empirically, yes. There is a later check for a non-zero mark, and if this line is omitted, the check doesn't fire the expected number of times and raises an internal error (with a small memory leak):
https://github.com/Rdatatable/data.table/blob/41e112348bb13d08132f50ab58b9d376c6343d07/src/assign.c#L882-L884

ben-schwen commented on src/hash.c:
```diff
@@ -45,38 +47,65 @@ static R_INLINE size_t get_full_size(size_t n_elements, double load_factor) {
+    return NULL; // #nocov
+
+  // Not strictly portable but faster than an explicit loop setting keys to NULL
+  struct hash_pair *table = calloc(n_full, sizeof(*table));
```
edit: I just tested all three options with `inline` and there is no noticeable different between `calloc`, `malloc` + `loop`, `malloc` + `memset`

MichaelChirico commented:
Probably also prudent to add something along the lines of "We've benchmarked the refactored code and find the performance satisfactory, but do report any edge case performance regressions we may have missed."

aitap commented:
Found the problem: changes to `marks` that happened inside `range_str` did not propagate to its caller. My bad. Sorry.

TysonStanley commented:
Really incredible work here. Looks like some conflicts to resolve before merging. Believe this is the last *necessary* merge before 1.18.0?

aitap commented:
Changes to GLCI config tested on GitLab:
https://gitlab.com/Rdatatable/data.table/-/pipelines/2214728801

UBSan report also addressed:
https://gitlab.com/Rdatatable/data.table/-/jobs/12435568788

We should probably specify the number of arguments when registering our
callMethods to avoid surprises like this from UBSan.
